packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,221 @@
+        const text = `$(${this.getStatusIcon(event.health)})`;
+        const tooltip = event.health ? `Connection health: ${event.health}%` : 'Not connected';
+        this.statusBar.setElement('connection-status', {
+            alignment: StatusBarAlignment.RIGHT,
+            text,
+            priority: 0,
+            tooltip
+        });
+    }
+
+    private getStatusIcon(health: number | undefined) {
+        if (health === undefined || health === 0) {
+            return 'exclamation-circle';
+        }
+        if (health < 25) {
+            return 'frown-o';
+        }
+        if (health < 50) {
+            return 'meh-o';
+        }
+        return 'smile-o';
+    }
+
+}
+
+export class ConnectionStateMachine implements ConnectionStatusChangeEvent {
+
+    static readonly MAX_HISTORY = 100;
+
+    public readonly health: number;
+
+    constructor(
+        private readonly props: { readonly threshold: number },
+        public readonly state: ConnectionState = ConnectionState.CONNECTED,
+        private readonly history: boolean[] = []) {
+
+        if (this.state === ConnectionState.CONNECTION_LOST) {
+            this.health = 0;
+        } else {
+            this.health = this.history.length === 0 ? 100 : Math.round((this.history.filter(success => success).length / this.history.length) * 100);
+        }
+    }
+
+    next(success: boolean): ConnectionStateMachine {
+        const newHistory = this.updateHistory(success);
+        // Initial optimism.
+        let hasConnection = true;
+        if (newHistory.length > this.props.threshold) {
+            hasConnection = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, hasConnection ? ConnectionState.CONNECTED : ConnectionState.CONNECTION_LOST, newHistory);
+    }
+
+    private updateHistory(success: boolean) {
+        const updated = [...this.history, success];
+        if (updated.length > ConnectionStateMachine.MAX_HISTORY) {
+            updated.shift();
+        }
+        return updated;
+    }
+
+}
+
+@injectable()
+export class ApplicationConnectionStatusContribution implements ConnectionStatusContribution {
+
+    private dialog: ConnectionStatusDialog | undefined;
+    private state = ConnectionState.CONNECTED;
+
+    constructor(
+        @inject(MessageService) protected readonly messageService: MessageService,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) { }
+
+    onStatusChange(event: ConnectionStatusChangeEvent): void {
+        if (this.state !== event.state) {
+            this.state = event.state;
+            switch (event.state) {
+                case ConnectionState.CONNECTION_LOST: {
+                    const message = 'Application connection to the Theia backend is lost. Performing retries...';
```

Comment:
`Theia appears to be offline. Attempting to reconnect...`?

Reply:
not sure about using Theia, since a concrete app can have another name, maybe `Your device` or `Your computer`

Reply:
You're right, I got rid of Theia.